### PR TITLE
feat: fallback email

### LINKED ISSUE #158: [Bug] Take email from other places
Currently, extension module takes email address from both `cart.billingAddress` and `cart.shippingAddress`. I think it would be great to add fallback logic and do:

```js
cart.billingAddress || cart.shippingAddress || cart.customerEmail || cart.customer?.email
```

About the third one (customerEmail) please make sure the field name is like that.

### PR DESCRIPTION
**Pull request by Fifciu**
## Description

closes #158 

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Testing

I've set `cart.customerEmail` field and I see the extension used this value.

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have updated the documentation where necessary
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works/doesn't break everything
- [x] Existing tests pass locally with my changes

### CODE REVIEW

File: extension/src/requestHandlers/createOrder.ts [hunk 1]
```diff
@@ -142,12 +142,13 @@ export function getCreateOrderParams(ctPayment: CTPayment, cart: CTCart): Promis
     const [paymentMethod, paymentIssuer] = ctPayment.paymentMethodInfo.method.split(',');
     const parsedCtPayment = JSON.parse(ctPayment.custom?.fields?.createPayment || '{}');
     const locale = parsedCtPayment.locale || configLocale;
+    const fallbackEmail = cart.customerEmail || cart.customer?.email;
```
Comment (denisand) [inline extension/src/requestHandlers/createOrder.ts:145]: Look below for a comment on types, first need to update the query to get the customer populated on cart (if possible on REST API)..

File: extension/src/types/index.ts [hunk 2]
```diff
@@ -90,6 +90,10 @@ export type CTCart = {
   billingAddress?: Object;
   locale?: string;
   shippingInfo?: Object;
+  customerEmail?: string;
+  customer?: {
```
[the remark this answers is not in the document]
Comment (Fifciu) [reply extension/src/types/index.ts:94]: Indeed. I need to fetch customer and read his email if cart has `customerId` and for that I will need a `view_customers` scope.

**Comment by Fifciu** [comment]
Ok. Updated. Now it requires `view_customers` scope too.

**Review by denisand — COMMENTED** [review]
One comment, the rest looks ok.
I ran it locally with different combinations of cart emails (from customerEmail, cart.email etc) and it worked well.
Need to add `view_customers` scope to required scopes in docs on https://github.com/mollie/commercetools/pull/162

File: extension/src/requestHandlers/createOrder.ts [hunk 3]
```diff
@@ -240,21 +241,37 @@ export default async function createOrder(
...
     const cartByPayment = await commercetoolsClient.execute(getCartByPaymentOptions);
     if (!cartByPayment.body.results.length) {
       const error = formatErrorResponse({ status: 404, message: `Could not find Cart associated with the payment ${paymentId}.` });
       return error;
     }
 
-    const orderParams = await getCreateOrderParams(ctPayment, cartByPayment.body.results[0]);
+    const cart = cartByPayment.body.results[0];
+    let customerEmail;
+    if (cart.customerId) {
```
[the remark this answers is not in the document]
Comment (Fifciu) [reply extension/src/requestHandlers/createOrder.ts]: You are totally right, thanks. I've just updated the PR.

**Review by lebradley — COMMENTED** [review]
Hey, thank you for the PR ! 

Please could you update the [scopes documentation](https://github.com/mollie/commercetools/blob/develop/docs/Installing_CommerceTools_APIExtension.md#commercetools-scopes) as well to add `view_customers`  ?

The pipeline is failing due to linting, you should be able to do a `npm run lint:fix` and it'll go through. 

Any questions, let me know 😄

File: extension/src/requestHandlers/createOrder.ts [hunk 4]
```diff
@@ -240,21 +241,39 @@ export default async function createOrder(
...
     const cartByPayment = await commercetoolsClient.execute(getCartByPaymentOptions);
     if (!cartByPayment.body.results.length) {
       const error = formatErrorResponse({ status: 404, message: `Could not find Cart associated with the payment ${paymentId}.` });
       return error;
     }
 
-    const orderParams = await getCreateOrderParams(ctPayment, cartByPayment.body.results[0]);
+    const cart = cartByPayment.body.results[0];
+    const missingShippingOrBillingEmail = (!cart.billingAddress.email || !cart.shippingAddress.email);
+    const shouldFetchCustomerEmail = !cart.customerEmail && missingShippingOrBillingEmail && cart.customerId;
```
[the remark this answers is not in the document]
Comment (Fifciu) [reply extension/src/requestHandlers/createOrder.ts:263]: I totally agree. Let's go for `(b)` way. So I think there is no need for additional code too. Am I right?

File: extension/src/requestHandlers/createOrder.ts [hunk 5]
```diff
@@ -240,21 +241,39 @@ export default async function createOrder(
...
+    let customerEmail;
+    if (shouldFetchCustomerEmail) {
```
[the remark this answers is not in the document]
Comment (Fifciu) [reply extension/src/requestHandlers/createOrder.ts:265]: Thanks!

File: extension/src/requestHandlers/createOrder.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (lebradley) [reply extension/src/requestHandlers/createOrder.ts:263]: Yep, let's go option B.

**Comment by lebradley** [comment]
Please could you update the [scopes documentation](https://github.com/mollie/commercetools/blob/develop/docs/Installing_CommerceTools_APIExtension.md#commercetools-scopes) as well to add view_customers ?

**Comment by Fifciu** [comment]
@lebradley Done

**Pull request merged**